### PR TITLE
add wrangle crop button to browse crops page

### LINKED ISSUE #620: Add crop wrangling link to the "browse crops" page
It would be handy to have a crop wrangling link at the top of the "browse crops" page if you're a wrangler.

### PR DESCRIPTION
**Pull request by robertlandreaux**
Fixes #620

### CODE REVIEW

**Comment by Skud** [comment]
So as you can see the tests for this failed!  Please run "rake" before submitting a pull request, and sort out any errors before sending it through.  In this case, the problem occurs when there's nobody signed in at all, because there is no such thing as current_member in that case.  You would have to test for the presence of current_member first.  However, there is a better way to do this, which is using CanCan's abilities, like so:

    if can? :wrangle, Crop

We should always use "can?" in preference to explicitly testing stuff around current_member, unless it can't be avoided (as was actually the case for a different thing I had to do yesterday, sigh.)

**Comment by tygriffin** [comment]
Sorry about the test thing. That's really my fault. I was showing Robert how to do the whole process while feeling a little tired myself.

**Comment by Skud** [comment]
Ah, no worries! It's easily fixed anyway.

(Also, a feature test to show that the button is there when it should be 
would be nice.)

On 17/01/2015 7:37 pm, Taylor Griffin wrote:
>
> Sorry about the test thing. That's really my fault. I was showing 
> Robert how to do the whole process while feeling a little tired myself.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/Growstuff/growstuff/pull/650#issuecomment-70359487>.
>

-- 
Alex "Skud" Bayley
skud@growstuff.org
http://growstuff.org/

**Comment by robertlandreaux** [comment]
I apologize, Alex. I will make sure to run the tests next time before I submit a pull request. Taylor and I did have a good pairing session besides this problem. I will look into the tests and consult Taylor for help. Thank you so much.

Robert

**Comment by tygriffin** [comment]
Oh and one more thing that I always forget: If you haven't already, add your name to the CONTRIBUTORS.md

**Comment by robertlandreaux** [comment]
@tygriffin @Skud Hello, everyone! The tests do not fail because of the prior issue; I think they do not run because there is a problem with the new test I wrote. Here is the error message I get on my machine: 

```growstuff/spec/features/crops/crop_wrangling_button_spec.rb:54:in `block (2 levels) in <top (required)>': undefined local variable or method `page' for RSpec::ExampleGroups::CropWranglingButton::CropWranglingButton_3:Class (NameError)```

I am fairly new to writing tests with Capybara, Factory Girl, and Rspec. I tried my best to write a feature test based upon my observation of other tests. I welcome any advice or criticism here and thank everybody in advance for helping me to contribute to this project. Contributing has been very fun and informative! I hope to be a good resource for Growstuff in the future.

Robert

File: spec/features/crops/crop_wrangling_button_spec.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+      background do
+        login_as(:member)
+      end
+
+      background do
+        visit crops_path
+      end
+
+      scenario "has no link to crop wrangling page"
```
Comment (tygriffin) [inline spec/features/crops/crop_wrangling_button_spec.rb]: You're missing a `do` here to open the block.

File: spec/features/crops/crop_wrangling_button_spec.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        expect(page).to have_no_link "Wrangle Crops", wrangle_crops_path
+      end
+    end
```
Comment (tygriffin) [inline spec/features/crops/crop_wrangling_button_spec.rb]: You need one more `end` here to close your top level feature block.

**Comment by tygriffin** [comment]
You have two problems that will prevent your tests from running - both related to the nesting of blocks. I've pointed them out inline. This is a real easy mistake to make so keep your eyes peeled!

File: spec/features/crops/crop_wrangling_button_spec.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
+require 'rails_helper'
+
+feature "crop wrangling button" do
+
+  let(:crop_wrangler_member) { FactoryGirl.create(:crop_wrangler_member) }
+
+    context "crop wrangling button" do
```
Comment (tygriffin) [inline spec/features/crops/crop_wrangling_button_spec.rb]: Good job using `context` to group tests, but probably only need to test two contexts. One when you're logged in as a wrangler and one when you're not. Testing the Admin member might be overkill (*in my opinion*).

File: spec/features/crops/crop_wrangling_button_spec.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+      background do
+        login_as(:crop_wrangler_member)
+      end
+
+      background do
+        visit crops_path
+      end
```
Comment (tygriffin) [inline spec/features/crops/crop_wrangling_button_spec.rb]: You can group these like so:
```
background do
  login_as(:crop_wrangler_member) # btw, you may one to double check the factory file where this is defined
  visit crops_path
end

File: spec/features/crops/crop_wrangling_button_spec.rb (hunk 3 above)
Comment (robertlandreaux) [inline spec/features/crops/crop_wrangling_button_spec.rb]: Excellent. I got rid of the Admin member part (lines 23-39).

File: spec/features/crops/crop_wrangling_button_spec.rb (hunk 4 above)
Comment (robertlandreaux) [inline spec/features/crops/crop_wrangling_button_spec.rb]: Got it. The spec/factories/member.rb file does say "crop_wrangling_member." Thank you for pointing me there!

File: spec/features/crops/crop_wrangling_button_spec.rb (hunk 2 above)
Comment (robertlandreaux) [inline spec/features/crops/crop_wrangling_button_spec.rb]: Got it.

File: spec/features/crops/crop_wrangling_button_spec.rb (hunk 1 above)
Comment (robertlandreaux) [inline spec/features/crops/crop_wrangling_button_spec.rb]: Got it.

**Comment by pozorvlak** [comment]
> Sorry about the test thing.

No worries - this is precisely why we use continuous integration :-)

File: app/views/crops/index.html.haml [hunk 5]
```diff
@@ -1,5 +1,6 @@
 - content_for :title, "Crops"
-
+  - if can? :wrangle, Crop
+  = link_to 'Wrangle Crops', wrangle_crops_path, :class => 'btn btn-primary'
```
Comment (pozorvlak) [inline app/views/crops/index.html.haml]: I think you need to indent this line one step further, so it's within the if-block.

File: spec/features/crops/crop_wrangling_button_spec.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+    context "crop wrangling button" do
+
+      background do
+        login_as(:crop_wrangler)
+        visit crops_path
+      end
+
+      scenario "has a link to crop wrangling page" do
+        expect(page).to have_link "Wrangle Crops", wrangle_crops_path
```
Comment (pozorvlak) [inline spec/features/crops/crop_wrangling_button_spec.rb]: The URL needs to be passed as a :href keyword parameter, like so:
```ruby
expect(page).to have_link "Wrangle Crops", :href => wrangle_crops_path
```

File: spec/features/crops/crop_wrangling_button_spec.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    context "crop wrangling button" do
+
+      background do
+        login_as(:member)
+        visit crops_path
+      end
+
+      scenario "has no link to crop wrangling page" do
+        expect(page).to have_no_link "Wrangle Crops", wrangle_crops_path
```
Comment (pozorvlak) [inline spec/features/crops/crop_wrangling_button_spec.rb]: As above, the URL should be passed as a keyword parameter.

**Comment by robertlandreaux** [comment]
@pozorvlak Thank you so much for your suggestions. Unfortunately, after applying them I am still getting a similar error:

```  1) crop wrangling button crop wrangling button has a link to crop wrangling page
     Failure/Error: expect(page).to have_link "Wrangle Crops", :href => wrangle_crops_path
       expected to find link "Wrangle Crops" with href "/crops/wrangle" but there were no matches
     # ./spec/features/crops/crop_wrangling_button_spec.rb:15:in `block (3 levels) in <top (required)>'```

Instead of pushing the code to my forked growstuff repository and thus updating the pull request, TravisCI, and Coveralls, would it be better to push the code somewhere else if I know I have a failing test, need help, and want to show my code to others? I am just wondering the best practice for Growstuff. I wouldn't want to negatively affect TravisCI or Coveralls with a failing test.

**Comment by jcaudle** [comment]
@robertlandreaux I've been woefully inactive on Growstuff for a while, but here are a few ways to do what you're hoping:

1. Commit and push on this branch in your fork. Travis will run a build and Coveralls will update, but you can always change what you commit and even change the commit history if you'd like later.
2. Checkout a new branch from where you are (`git checkout -b some-other-branch-you-arent-using-elsewhere`) and commit your changes there, then push that and point people at it. The link for this branch in a compare view, for example is https://github.com/robertlandreaux/growstuff/compare/add-crop-wrangling-link. Just change out the branch name and you'll be set. You might want to preface the branch name with `spike/` to indicate that you don't want it to be merged into `dev`.
3. Open a [Gist](gist.github.com) with the relevant code that you want someone to take a peek at.

I hope this helps.

**Comment by pozorvlak** [comment]
I think the best thing is to commit and push on this fork - that way we
have easy access to the code and the test results. Don't worry about
negatively affecting our Travis scores - only builds of the main branches
show up on the project page. If you look at our [build history](
https://travis-ci.org/Growstuff/growstuff/builds) you'll see that I've been
responsible for five failing builds in the last 24 hours, so it really
isn't a big deal :-)

**Comment by robertlandreaux** [comment]
@jcaudle @pozorvlak Thank you all so much for your suggestions! Joseph, I am really happy to know about your options 2 and 3. I will push my code to this pull request again.

**Comment by Skud** [comment]
What's the status of this PR?

**Comment by robertlandreaux** [comment]
Taylor and I are pairing up Saturday, February 7th. He was going to help me finish up the crop_wrangling_button_spec.rb.

**Comment by Skud** [comment]
Thanks!

**Pull request merged**